L1Trigger/L1TZDC/plugins/L1TZDCProducer.cc

```diff
@@ -0,0 +1,237 @@
+      unsigned short adc = (unsigned short)frame[iTS].adc();
+      rawadc[ihitid][iTS] = adc;
+    }  // end of loop over iTS
+  }    //end of loop over channels
+
+  for (int ibx = 0; ibx < nSamples; ibx++) {
+    double cEMP = 0, cEMM = 0, cHDP = 0, cHDM = 0;
+    double sumcEMP = 0, sumcEMM = 0, sumcHDP = 0, sumcHDM = 0;
+    //idet=0-4 correpond to the EM channels
+    for (int idet = 0; idet < 5; idet++) {
+      unsigned short EMP = rawadc[idet + 9][ibx];
+      unsigned short EMM = rawadc[idet][ibx];
+
+      int cEMP_LUTIndex = zdcLUTIndexHelper(idet + 9, (int)EMP);
+      int cEMM_LUTIndex = zdcLUTIndexHelper(idet, (int)EMM);
+
+      cEMP = ((double)m_params->zdcLUT()->data(cEMP_LUTIndex)) / ((double)m_scaleFactor);
+      cEMM = ((double)m_params->zdcLUT()->data(cEMM_LUTIndex)) / ((double)m_scaleFactor);
+
+      sumcEMP = sumcEMP + cEMP;
+      sumcEMM = sumcEMM + cEMM;
+    }
+    //idet=5-8 correspond to HAD channels
+    for (int idet = 5; idet < 9; idet++) {
+      unsigned short HDP = rawadc[idet + 9][ibx];
+      unsigned short HDM = rawadc[idet][ibx];
+
+      int cHDP_LUTIndex = zdcLUTIndexHelper(idet + 9, (int)HDP);
+      int cHDM_LUTIndex = zdcLUTIndexHelper(idet, (int)HDM);
+
+      cHDP = ((double)m_params->zdcLUT()->data(cHDP_LUTIndex)) / ((double)m_scaleFactor);
+      cHDM = ((double)m_params->zdcLUT()->data(cHDM_LUTIndex)) / ((double)m_scaleFactor);
+
+      sumcHDP = sumcHDP + cHDP;
+      sumcHDM = sumcHDM + cHDM;
+    }
+    double sumM = sumcEMM + sumcHDM;
+    double sumP = sumcEMP + sumcHDP;
+
+    if (ibx == 4) {
+      edm::LogInfo("L1TZDCProducer") << ", sumM= " << sumM << std::endl;
+      edm::LogInfo("L1TZDCProducer") << ", sumP= " << sumP << std::endl;
+    }
+    l1t::EtSum tempEtM = l1t::EtSum();
+    tempEtM.setHwPt(sumM);
+    tempEtM.setHwEta(-1.);
+    tempEtM.setHwPhi(0.);
+    tempEtM.setType(EtSum::EtSumType::kZDCM);
+
+    l1t::EtSum tempEtP = l1t::EtSum();
+    tempEtP.setHwPt(sumP);
+    tempEtP.setHwEta(1.);
+    tempEtP.setHwPhi(0.);
+    tempEtP.setType(EtSum::EtSumType::kZDCP);
+
+    //One object distinguished by type P, M
+    etsums.push_back(ibx, CaloTools::etSumP4Demux(tempEtP));
+    etsums.push_back(ibx, CaloTools::etSumP4Demux(tempEtM));
+
+    if (ibx >= startBXR && ibx <= endBXR) {
+      etsumsReduced.push_back(ibx - startBXR, CaloTools::etSumP4Demux(tempEtP));
+      etsumsReduced.push_back(ibx - startBXR, CaloTools::etSumP4Demux(tempEtM));
+    }
+  }  // end of loop over bunch crossings
+
+  iEvent.emplace(m_etToken, std::move(etsumsReduced));
+}
+
+// ------------ method called when starting to processes a run  ------------
+void L1TZDCProducer::beginRun(edm::Run const& iRun, edm::EventSetup const& iSetup) {
+  edm::ESHandle<CaloParams> candidateHandle = iSetup.getHandle(m_candidateToken);
+
+  m_params = std::make_unique<CaloParamsHelper>(*candidateHandle.product());
+  m_scaleFactor = m_params->zdcLUT()->data(0);  //First position is the integer scaling factor
```

Comment:
On a longer term I would suggest to write an `ESProducer` that can produce directly the `CaloParamsHelper` object.

Then the `m_params` data member and this `beginRun` method could be removed, and the EventSetup payload used directly.